### PR TITLE
Change `needsUpdate` to a `state` enum and use that for better glob invalidation

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/build/issues/622

### CODE REVIEW

File: build_runner/lib/src/asset_graph/node.dart [hunk 1]
```diff
@@ -105,8 +116,8 @@ class GeneratedAssetNode extends AssetNode {
   /// The primary input which generated this node.
   final AssetId primaryInput;
 
-  /// Whether or not this asset needs to be updated.
-  bool needsUpdate;
+  /// The current [GeneratedNodeState] for this node.
```
Comment (natebosch) [inline build_runner/lib/src/asset_graph/node.dart]: drop the comment?

File: build_runner/lib/src/generate/build_impl.dart [hunk 2]
```diff
@@ -469,20 +469,23 @@ class _SingleBuild {
...
     // We only check the first output, because all outputs share the same inputs
     // and invalidation state.
-    if (!node.needsUpdate) return false;
+    if (node.state == GeneratedNodeState.upToDate) return false;
+    // Early bail out condition, this is a forced update.
+    if (node.state == GeneratedNodeState.definitelyNeedsUpdate) return true;
     // TODO: Don't assume the worst for globs
     // https://github.com/dart-lang/build/issues/624
-    if (node.previousInputsDigest == null || node.globs.isNotEmpty) {
+    if (node.previousInputsDigest == null) {
```
Comment (natebosch) [inline build_runner/lib/src/generate/build_impl.dart]: ah, I guess we could have also considered clearing out the previous digest to indicate a glob change...

I think I like the enum better.

File: build_runner/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## 0.7.12-prerelease
```
Comment (natebosch) [inline build_runner/CHANGELOG.md]: is "prerelease" better than "dev" in some way?

Comment (jakemac53) [reply build_runner/CHANGELOG.md]: I think so yes, dev released in theory could be published, this more clearly indicates that it's not imo.

File: build_runner/lib/src/asset_graph/node.dart (hunk 1 above)
Comment (jakemac53) [reply build_runner/lib/src/asset_graph/node.dart]: done

File: build_runner/lib/src/generate/build_impl.dart (hunk 2 above)
Comment (jakemac53) [reply build_runner/lib/src/generate/build_impl.dart]: ack

**Comment by jakemac53** [comment]
Merged on red because the hanging tests look related to a package dependency or sdk update (they fail on master as well)

**Pull request merged**